### ISSUE TITLE
Inconsistency of square error between it's estimation during training and post evaluation

### ISSUE THREAD

**Issue by snurkabill** [Microsoft/CNTK#1381]
Hello! 

I've noticed an inconsitency between training CNTK squared error and post evaluation CNTK squared error. I will describe our use-case in detail: 

We run so-called stages. Each stage contains some CNTK epochs (so each stage CNTK training is executed on already existing model (except first epoch of first stage)). Now I will post graph we are getting. You can see that around 150th stage, the estimated error[1] of first epoch in CNTK training starts to diverge from post eval error[2] calculation. I am not saying this is bug, but we don't understand such behaviour. For completeness, I am adding training and stopping(validation) error, so we can see that this this "weird" error values holds  between different parts of dataset. One can see negative corelation between yellow and blue/red line - which is certainly not expected. 

![image](https://cloud.githubusercontent.com/assets/4818573/22491311/d46c5c06-e822-11e6-9b5f-2c98503c6947.png)

[1] The estimated error of first epoch in stage looks like this: 
Finished Epoch[ 1 of 10]: [Training] _lossFunction = 0.54939474 * 27407; _errors = 52.716% * 27407; totalSamplesSeen = 27407; learningRatePerSample = 0.00078125001; epochTime=5.26347s

[2] The post evaluation is just running CNTK "eval" phase on static data with no training.

**Comment by cha-zhang** [comment]
Thanks for reporting this. Is there a simple sample that we can repro this issue?

**Comment by snurkabill** [comment]
I am going to try reproduce it on CNTK example datasets (with my setup of heursitics). if it won't be reproducible, I will try extract any data from business-sensitive info and try to create reproducer from it.

**Comment by snurkabill** [comment]
Still working on standelone reproducer. 

I have question and theory: 
question: 
When we see line like this one: "Finished Epoch[ 1 of 10]: [Training] _lossFunction = 0.54939474 * 27407; _errors = 52.716% * 27407; totalSamplesSeen = 27407; " was the _lossFunction calculated from final(non-chaning) model for epoch 1 or was it estimated(averaged) over each mini-batch while model was changing each batch? Is it calculated when dropout is applied? We've checked the code and it seems that is it estimated(averaged) error, I am asking to be sure. 

theory: 
consider neural network with several hidden layers and linear output layer trained with dropou 0.5 rate. Is it possible, that when even dropouted network is overfitted, then the ealuation on network with all connections is greater? I will demonstrate it on example: 

consider network N with weights W. Now consider two disjunctive subsets of weights W, W1 and W2. We can think about two neural networks N1 and N2 with weights W1 and W2 respectively (this might be done with lucky dropout). 

Now consider tuple (x, y) where y = (1, 0). When W1 and W2 are overfitted, both of them result in vector close to y, when x is on input. What happens, when we evaluate N on x? Both subnetworks end up with let's say (0.999, 0.001), so at linear output layer, they are just added together, which is vector (1.998, 0.001) so the error high. Is it possible or am I missing something?

**Comment by cha-zhang** [comment]
The error reported during training is an estimate. It is calculated when dropout is applied. 

One reason that dropout is very robust is because nodes are left out randomly. You theory assumes a fixed set of nodes are set to zero, which is not the case. 

I can't tell why something like your curve could happen. Likely there is some overfitting happening (yes, overfitting can still happen with dropout). Try to use a higher dropout rate, or smaller number of hidden nodes, or adding more other regularizations (weight decay, L1, etc.).

**Comment by jakrivan** [comment]
Thanks cha-zhang for the info.

> It is calculated when dropout is applied.
To be absolutely sure - you mean _after_ the dropout is applied, correct?

My colleague was able to create standalone repro case demonstrating this behavior (occuring only when repeatedly resuming from last checkpoint - so we suspect that might be the culprit). We'll try to reach out offline with it as our company policy is strict on posting our code/data on publicly accessible forums.

**Comment by cha-zhang** [comment]
The issue has been identified. Repeating the same set of epochs is a bad idea because CNTK's random seeds depend on epoch number, thus always training from epoch 1-10 has the risk of letting the model learn the random pattern of these fixed set of seeds, causing divergence in training and testing error.

**Comment by tangyuq** [comment]
As the issue is identified as unusual specific training/data feeding procedure related, I am closing this issue for now. Please feel free to re-open this issue if you have further comments.
